File: src/1133_largest_unique_number/largest_unique_number.py
from typing import Counter


class Solution:
    def largestUniqueNumber(self, nums: list[int]) -> int:
        counts = Counter(nums)

        uniqueNums = [num for num, freq in counts.items() if freq == 1]
        return max(uniqueNums) if uniqueNums else -1


# Test cases
def test_solution():
    solution = Solution()

    # Test case 1
    nums1 = [5, 7, 3, 9, 4, 9, 8, 3, 1]
    result1 = solution.largestUniqueNumber(nums1)
    expected1 = 8
    print(f"{result1} | {expected1} | {'PASS' if result1 == expected1 else 'FAIL'}")

    # Test case 2
    nums2 = [9, 9, 8, 8]
    result2 = solution.largestUniqueNumber(nums2)
    expected2 = -1
    print(f"{result2} | {expected2} | {'PASS' if result2 == expected2 else 'FAIL'}")

    # Test case 3
    nums3 = [1, 1, 1, 1, 2, 2, 2, 3]
    result3 = solution.largestUniqueNumber(nums3)
    expected3 = 3
    print(f"{result3} | {expected3} | {'PASS' if result3 == expected3 else 'FAIL'}")

    # Test case 4
    nums4 = [5]
    result4 = solution.largestUniqueNumber(nums4)
    expected4 = 5
    print(f"{result4} | {expected4} | {'PASS' if result4 == expected4 else 'FAIL'}")

    # Test case 5
    nums5 = [0, 1, 0, 2, 3, 3]
    result5 = solution.largestUniqueNumber(nums5)
    expected5 = 2
    print(f"{result5} | {expected5} | {'PASS' if result5 == expected5 else 'FAIL'}")


if __name__ == "__main__":
    test_solution()

File: src/1133_largest_unique_number/largest_unique_number.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
   public:
    int largestUniqueNumber(vector<int>& nums) {
        unordered_map<int, int> counts;

        for (int num : nums) {
            counts[num] += 1;
        }

        int maxNumber = -1;

        for (const auto& [key, value] : counts) {
            if (value == 1) maxNumber = max(maxNumber, key);
        }

        return maxNumber;
    }
};

// Helper function for testing
void testSolution() {
    Solution solution;

    // Test case 1
    vector<int> nums1 = {5, 7, 3, 9, 4, 9, 8, 3, 1};
    int result1 = solution.largestUniqueNumber(nums1);
    int expected1 = 8;
    cout << "Test 1: " << (result1 == expected1 ? "PASS" : "FAIL") << endl;

    // Test case 2
    vector<int> nums2 = {9, 9, 8, 8};
    int result2 = solution.largestUniqueNumber(nums2);
    int expected2 = -1;
    cout << "Test 2: " << (result2 == expected2 ? "PASS" : "FAIL") << endl;

    // Test case 3
    vector<int> nums3 = {1, 1, 1, 1, 2, 2, 2, 3};
    int result3 = solution.largestUniqueNumber(nums3);
    int expected3 = 3;
    cout << "Test 3: " << (result3 == expected3 ? "PASS" : "FAIL") << endl;

    // Test case 4
    vector<int> nums4 = {5};
    int result4 = solution.largestUniqueNumber(nums4);
    int expected4 = 5;
    cout << "Test 4: " << (result4 == expected4 ? "PASS" : "FAIL") << endl;

    // Test case 5
    vector<int> nums5 = {0, 1, 0, 2, 3, 3};
    int result5 = solution.largestUniqueNumber(nums5);
    int expected5 = 2;
    cout << "Test 5: " << (result5 == expected5 ? "PASS" : "FAIL") << endl;
}

int main() {
    testSolution();
    return 0;
}
